paperqa/agents/search.py

```diff
@@ -326,19 +333,24 @@ async def _add_document() -> None:
             )
             raise
 
-    @staticmethod
+    async def checkpoint(self) -> None:
+        """Commit all pending changes, merge threads, and reset writer, as writers are not valid after commit."""
+        writer: IndexWriter = await self.writer
+        writer.commit()
+        writer.wait_merging_threads()
+        self._writer = None
```

Comment:
This pattern here where you have:

- Handling of `_writer` in `self.writer`
- Use the public `self.writer` here
- Also assign `_writer`

It's awkward and breaks encapsulation. Can you revisit this logic a bit? Try to think of ways you can avoid manipulating both private and public stuff here

Reply:
We can do 
```python
        self._writer.commit()
        self._writer.wait_merging_threads()
        self._writer = None
```
As there has to be an active writer to call this method.

Reply:
The only problem is that the pre-commit will fail because `self._writer` can be have type `None`